src/main/scala/tech/cryptonomic/conseil/tezos/TezosNodeMultiFetch.scala

```diff
@@ -0,0 +1,164 @@
+  * Thus the required implementation is in the form of a `cats.data.Kleisli` type,
+  * parametric on effect `F`, input `In`, output `Out`. As needed, the in/out types are
+  * wrapped in a `T[_]` type constructor that represents multiple values (e.g. a collection type).
+  *
+  * @tparam F the higher-order effect type (e.g. `Try`, `Future`, `IO`, `Either`)
+  * @tparam T a traversable type (used to wrap mutliple requests and corresponding return values)
+  */
+trait TezosNodeMultiFetch[F[_], T[_]] {
+  /** the input type*/
+  type In
+  /** the output type*/
+  type Out
+  /** the json representation type used*/
+  type JsonType
+
+  /** an effectful function from a collection of inputs `T[In]`
+    * to the collection of json, tupled with the corresponding input `T[(In, JsonType)]`
+    */
+  def fetchJsonBatch: Kleisli[F, T[In], T[(In, JsonType)]]
+
+  /** an effectful function that decodes the json value to an output `Out` */
+  def decodeJson: Kleisli[F, JsonType, Out]
+
+  /** using the combination of the provided functions we fetch all data with this general outline:
+    * 1. get json values from the node, given a traversable input collection
+    * 2. keep the correlation with the input and convert every json to a corresponding output
+    * 3. return the traversable collection containing paired input and output values, all wrapped in the effect F
+    */
+  def fetch(implicit app: Monad[F], tr: Traverse[T]): Kleisli[F, T[In], T[(In, Out)]] =
+    fetchJsonBatch.andThen(json => decodeJson.second[In].traverse(json))
+
+}
+
+object TezosNodeMultiFetch {
+
+  /* An alias used to define function constraints on the internal dependent types (i.e. `I`, `O`, `J`)
+   * which would be otherwise transparent in any function signature
+   */
+  private type Aux[F[_], T[_], I, O, J] = TezosNodeMultiFetch[F, T] {
+      type In = I
+      type Out = O
+      type JsonType = J
+    }
+
+  /** Adds an extra json-decoding operation for every json intermediate result, returning both the
+    * original output `O` and the new decoded output `O2` as a tuple
+    */
+  def addDecoding[F[_]: Apply, T[_], I, O, O2, J](multiFetch: Aux[F,T,I,O,J], additionalDecode: Kleisli[F, J, O2]) = new TezosNodeMultiFetch[F, T] {
+    type In = I
+    type Out = (O, O2)
+    type JsonType = J
+
+    def fetchJsonBatch = multiFetch.fetchJsonBatch
+
+    def decodeJson = multiFetch.decodeJson.product(additionalDecode)
+  }
+
+  /** Allows to add an additional decoder to a "multi-fetch" instance, by providing an extension method `decodeAlso(additionalDecode)`,
+    * subject to `F` having a `cats.Apply` instance
+    */
+  object Syntax {
+    implicit class MultiFetchOps[F[_], T[_], I, O, J](m: Aux[F, T, I, O, J]) {
+      def decodeAlso[O2](additionalDecode: Kleisli[F, J, O2])(implicit app: Apply[F]): Aux[F, T, I, (O, O2), J] = addDecoding(m, additionalDecode)
+    }
+  }
+
+}
+
+/** Defines intances of `TezosNodeMultiFetch` for block-related data */
```

Comment:
Typo in description. Also, maybe it would make more sense to split up the interface from the actual implementation? In general, the Tezos folder is getting bloated. One option is to have sub folders for database operations, api operations, node operations, and abstractions/interfaces, and place concrete implementations of abstractions in the folder in which those concrete implementations are used.

Reply:
I did it, I think we mentioned this already somewhere